opencypher/tests/ast/test_with.py

```diff
@@ -0,0 +1,81 @@
+from hamcrest import assert_that, equal_to, is_
+from parameterized import parameterized
+
+from opencypher.ast import (
+    Expression,
+    ExpressionAlias,
+    Limit,
+    NonEmptySequence,
+    Order,
+    ReturnBody,
+    ReturnItem,
+    Skip,
+    SortItem,
+    With,
+    Where,
+)
+
+
+@parameterized([
+    (
+        False, ["foo", "bar", "baz"], None, None, None, None,
+        "WITH foo, bar AS baz",
+    ),
+    (
+        False, ["()", "foo"], None, None, None, None,
+        "WITH () AS foo",
+    ),
+    (
+        True, ["()"], None, None, None, None,
+        "WITH DISTINCT ()",
+    ),
+    (
+        True, ["()", "foo"], None, None, None, None,
+        "WITH DISTINCT () AS foo",
+    ),
+    (
+        True, ["()"], None, None, None, "Bar",
+        "WITH DISTINCT () WHERE Bar",
+    ),
+    (
+        True, ["()", "Foo"], None, None, None, "Bar",
+        "WITH DISTINCT () AS Foo WHERE Bar",
+    ),
+    (
+        False, ["otherPerson", "foaf"], None, None, None, "foaf > 1",
+        "WITH otherPerson AS foaf WHERE foaf > 1",
+    ),
+    (
+        False, ["otherPerson", "count(*)", "foaf"], None, None, None, "foaf > 1",
+        "WITH otherPerson, count(*) AS foaf WHERE foaf > 1",
+    ),
+    (
+        False, ["n"], "n.name DESC", None, 3, None,
+        "WITH n ORDER BY n.name DESC LIMIT 3",
+    ),
+])
```

Comment:
Should I add a test for multiple WITHs? How do I do this?

Reply:
You'd have to extend your parameterization. The normal way I'd do this is to break out separate unit tests for each unit of the AST: 
 - Write one parameterized test for `With`, validating the various combinations of arguments to it.
 - Write one parameterized test for `WithSubQuery`, validating various *numbers* of arguments to it, but don't bother with varying the arguments to *its* arguments.
 - Write one parameterized test for `MultiPartQuery`, similarly focusing on number and nullity of *its* arguments vs the specifics of what each argument is.

You're essentially write tests that are provably transitive as long as the type system works properly. You don't have to test every argument of `With` at the same time you test arguments of `WithSubQuery`.